File: clients/typescript/src/channels/state.ts
import { ServerError } from "../errors";
import { makeMsg } from "../protocol";
import {
  AliasInfo, CHAN_STATE, FuncInfo, InspectItem, Message, VarInfo,
} from "../types";

type SendFn = (msg: Message) => Promise<void>;
type RecvFn = (ch: number) => Promise<Message>;

export class StateChannel {
  constructor(private send: SendFn, private recv: RecvFn) {}

  private async request(msg: Message): Promise<Message> {
    await this.send(msg);
    const resp = await this.recv(CHAN_STATE);
    if (resp.type === "error") {
      throw new ServerError(String(resp.message ?? "state error"), CHAN_STATE);
    }
    return resp;
  }

  // Variables
  async getVar(name: string): Promise<VarInfo> {
    const resp = await this.request(makeMsg(CHAN_STATE, "get", { target: "var", name }));
    return {
      name: String(resp.name ?? name),
      value: String(resp.value ?? ""),
      attributes: (resp.attributes as string[]) ?? [],
    };
  }

  async setVar(name: string, value: string, attributes?: string[]): Promise<void> {
    const extra: Record<string, unknown> = { target: "var", name, value };
    if (attributes) extra.attributes = attributes;
    await this.request(makeMsg(CHAN_STATE, "set", extra));
  }

  async unsetVar(name: string): Promise<void> {
    await this.request(makeMsg(CHAN_STATE, "unset", { target: "var", name }));
  }

  // Functions
  async getFunc(name: string): Promise<FuncInfo> {
    const resp = await this.request(makeMsg(CHAN_STATE, "get", { target: "function", name }));
    return { name: String(resp.name ?? name), definition: String(resp.value ?? "") };
  }

  async unsetFunc(name: string): Promise<void> {
    await this.request(makeMsg(CHAN_STATE, "unset", { target: "function", name }));
  }

  // Aliases
  async getAlias(name: string): Promise<AliasInfo> {
    const resp = await this.request(makeMsg(CHAN_STATE, "get", { target: "alias", name }));
    return { name: String(resp.name ?? name), value: String(resp.value ?? "") };
  }

  async setAlias(name: string, value: string): Promise<void> {
    await this.request(makeMsg(CHAN_STATE, "set", { target: "alias", name, value }));
  }

  async unsetAlias(name: string): Promise<void> {
    await this.request(makeMsg(CHAN_STATE, "unset", { target: "alias", name }));
  }

  // Traps
  async setTrap(signal: string, command: string): Promise<void> {
    await this.request(makeMsg(CHAN_STATE, "set", { target: "trap", name: signal, value: command }));
  }

  async unsetTrap(signal: string): Promise<void> {
    await this.request(makeMsg(CHAN_STATE, "unset", { target: "trap", name: signal }));
  }

  // Inspect
  async inspect(query: string): Promise<InspectItem[]> {
    const resp = await this.request(makeMsg(CHAN_STATE, "inspect", { query }));
    return (resp.data as InspectItem[]) ?? [];
  }
}

File: clients/go/bashclient/state.go
package bashclient

import "context"

// StateChannel handles channel 2: variable, function, alias, and trap management.
type StateChannel struct {
	send sendFunc
	recv recvFunc
}

func newStateChannel(send sendFunc, recv recvFunc) *StateChannel {
	return &StateChannel{send: send, recv: recv}
}

func (c *StateChannel) request(ctx context.Context, msg Message) (Message, error) {
	if err := c.send(msg); err != nil {
		return nil, err
	}
	resp, err := c.recv(ctx, ChanState)
	if err != nil {
		return nil, err
	}
	if getString(resp, "type") == "error" {
		return nil, newServerError(getString(resp, "message"), ChanState)
	}
	return resp, nil
}

// GetVar retrieves a variable.
func (c *StateChannel) GetVar(ctx context.Context, name string) (*VarInfo, error) {
	resp, err := c.request(ctx, MakeMsg(ChanState, "get", map[string]any{
		"target": "var", "name": name,
	}))
	if err != nil {
		return nil, err
	}
	return &VarInfo{
		Name:       getString(resp, "name"),
		Value:      getString(resp, "value"),
		Attributes: getStringSlice(resp, "attributes"),
	}, nil
}

// SetVar sets a variable.
func (c *StateChannel) SetVar(ctx context.Context, name, value string, attributes []string) error {
	msg := MakeMsg(ChanState, "set", map[string]any{
		"target": "var", "name": name, "value": value,
	})
	if len(attributes) > 0 {
		msg["attributes"] = attributes
	}
	_, err := c.request(ctx, msg)
	return err
}

// UnsetVar removes a variable.
func (c *StateChannel) UnsetVar(ctx context.Context, name string) error {
	_, err := c.request(ctx, MakeMsg(ChanState, "unset", map[string]any{
		"target": "var", "name": name,
	}))
	return err
}

// GetFunc retrieves a function definition.
func (c *StateChannel) GetFunc(ctx context.Context, name string) (*FuncInfo, error) {
	resp, err := c.request(ctx, MakeMsg(ChanState, "get", map[string]any{
		"target": "function", "name": name,
	}))
	if err != nil {
		return nil, err
	}
	return &FuncInfo{
		Name:       getString(resp, "name"),
		Definition: getString(resp, "value"),
	}, nil
}

// UnsetFunc removes a function.
func (c *StateChannel) UnsetFunc(ctx context.Context, name string) error {
	_, err := c.request(ctx, MakeMsg(ChanState, "unset", map[string]any{
		"target": "function", "name": name,
	}))
	return err
}

// GetAlias retrieves an alias.
func (c *StateChannel) GetAlias(ctx context.Context, name string) (*AliasInfo, error) {
	resp, err := c.request(ctx, MakeMsg(ChanState, "get", map[string]any{
		"target": "alias", "name": name,
	}))
	if err != nil {
		return nil, err
	}
	return &AliasInfo{
		Name:  getString(resp, "name"),
		Value: getString(resp, "value"),
	}, nil
}

// SetAlias sets an alias.
func (c *StateChannel) SetAlias(ctx context.Context, name, value string) error {
	_, err := c.request(ctx, MakeMsg(ChanState, "set", map[string]any{
		"target": "alias", "name": name, "value": value,
	}))
	return err
}

// UnsetAlias removes an alias.
func (c *StateChannel) UnsetAlias(ctx context.Context, name string) error {
	_, err := c.request(ctx, MakeMsg(ChanState, "unset", map[string]any{
		"target": "alias", "name": name,
	}))
	return err
}

// SetTrap sets a trap handler.
func (c *StateChannel) SetTrap(ctx context.Context, signal, command string) error {
	_, err := c.request(ctx, MakeMsg(ChanState, "set", map[string]any{
		"target": "trap", "name": signal, "value": command,
	}))
	return err
}

// UnsetTrap removes a trap handler.
func (c *StateChannel) UnsetTrap(ctx context.Context, signal string) error {
	_, err := c.request(ctx, MakeMsg(ChanState, "unset", map[string]any{
		"target": "trap", "name": signal,
	}))
	return err
}

// Inspect queries state information.
func (c *StateChannel) Inspect(ctx context.Context, query string) ([]Message, error) {
	resp, err := c.request(ctx, MakeMsg(ChanState, "inspect", map[string]any{
		"query": query,
	}))
	if err != nil {
		return nil, err
	}
	items := getSlice(resp, "data")
	result := make([]Message, 0, len(items))
	for _, item := range items {
		if m, ok := item.(map[string]any); ok {
			result = append(result, Message(m))
		}
	}
	return result, nil
}
